src/plugins/expressions/common/expression_functions/specs/map_column.ts

```diff
@@ -80,50 +80,50 @@ export const mapColumn: ExpressionFunctionDefinition<
       default: null,
     },
   },
-  fn: (input, args) => {
-    const expression = (...params: Parameters<Required<MapColumnArguments>['expression']>) =>
-      args
-        .expression?.(...params)
-        .pipe(take(1))
-        .toPromise() ?? Promise.resolve(null);
-    const columnId = args.id != null ? args.id : args.name;
+  fn(input, args) {
+    const id = args.id ?? args.name;
 
-    const columns = [...input.columns];
-    const rowPromises = input.rows.map((row) => {
-      return expression({
-        type: 'datatable',
-        columns,
-        rows: [row],
-      }).then((val) => ({
-        ...row,
-        [columnId]: val,
-      }));
-    });
+    return defer(() => {
+      const rows$ = input.rows.length
+        ? zip(
+            ...input.rows.map((row) =>
+              (
+                args.expression?.({
+                  type: 'datatable',
+                  columns: [...input.columns],
+                  rows: [row],
+                }) ?? of(null)
+              ).pipe(map((value) => ({ ...row, [id]: value })))
+            )
+          )
+        : of([]);
 
-    return Promise.all(rowPromises).then((rows) => {
-      const existingColumnIndex = columns.findIndex(({ name }) => name === args.name);
-      const type = rows.length ? getType(rows[0][columnId]) : 'null';
-      const newColumn = {
-        id: columnId,
-        name: args.name,
-        meta: { type },
-      };
-      if (args.copyMetaFrom) {
-        const metaSourceFrom = columns.find(({ id }) => id === args.copyMetaFrom);
-        newColumn.meta = { ...newColumn.meta, ...(metaSourceFrom?.meta || {}) };
-      }
+      return rows$.pipe<Datatable>(
```

Comment:
why do we need to pipe here?

Reply:
That's here because we need to process rows altogether and put them into a `datatable`.